### PR TITLE
Define zip on Option

### PR DESCRIPTION
**Pull request by dwijnand**
Fixes scala/bug#8394

### CODE REVIEW

File: test/junit/scala/OptionTest.scala [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package scala
+
+import org.junit.Assert._
+import org.junit.Test
+
+class OptionTest {
+  @Test def testOptionZip1: Unit = assertEquals(Some("foo") zip List("bar"), Some(("foo", "bar")))
+  @Test def testOptionZip2: Unit = assertEquals(Some("foo") zip Nil, None)
+  @Test def testOptionZip3: Unit = assertEquals(None zip List("bar"), None)
```
Comment (viktorklang) [inline test/junit/scala/OptionTest.scala]: A test case with a collection of size > 1 as well?

Comment (dwijnand) [reply test/junit/scala/OptionTest.scala]: good idea

**Comment by julienrf** [comment]
Would it be too breaking to take an `Option` as parameter rather than an `Iterable`? I find it confusing. Will the existing implicit conversion be triggered when called with a collection?

**Comment by dwijnand** [comment]
> Would it be too breaking to take an `Option` as parameter rather than an `Iterable`?

good question, I'm not sure. perhaps we should run that change against the community build? (Seth?)

> Will the existing implicit conversion be triggered when called with a collection?

the existing implicit conversion is from option to collection, not the other way round, unless I've misunderstood you..

**Comment by julienrf** [comment]
I meant, in case we have `def zip[B](that: Option[B]): Option[(A, B)]`, then users already using `maybeFoo.zip(list)` will get a compilation error. Unless the implicit conversion from `Option` to collection is triggered, in which case the existing code will always compile.

**Comment by dwijnand** [comment]
> Unless the implicit conversion from `Option` to collection is triggered, in which case the existing code will always compile.

if zip on Option takes Option, then no amount of Option to collection conversion triggering is going to help.

here's what the test looks like if you change the signature:

```scala
[error] /d/scala/test/junit/scala/OptionTest.scala:7: type mismatch;
[error]  found   : List[String]
[error]  required: Option[?]
[error]   @Test def testZip1: Unit = assertEquals(Some("foo") zip List("bar", "baz"), Some(("foo", "bar")))
[error]                                                               ^
[error] /d/scala/test/junit/scala/OptionTest.scala:8: type mismatch;
[error]  found   : scala.collection.immutable.Nil.type
[error]  required: Option[?]
[error]   @Test def testZip2: Unit = assertEquals(Some("foo") zip Nil, None)
[error]                                                           ^
[error] /d/scala/test/junit/scala/OptionTest.scala:9: type mismatch;
[error]  found   : List[String]
[error]  required: Option[?]
[error]   @Test def testZip3: Unit = assertEquals(None zip List("bar"), None)
[error]                                                        ^
[error] three errors found
```

**Comment by hrhino** [comment]
> Would it be too breaking to take an Option as parameter rather than an Iterable? I find it confusing. 

👍. I think being able to say `option.zip(list)` is a misfeature which could lead to surprising bugs. It's the same as `option.zip(list.headOption)` which I wouldn't consider overly burdensome, as compared to the potential of thinking one or both sides are full collections (not `Option`s).

Since 2.13 is going to require users to adjust their collections usage anyways, I propose that we can drop this feature. (I do wonder how often it's used in the wild; I count zero uses in my code FWLIW).

**Comment by julienrf** [comment]
> if zip on Option takes Option, then no amount of Option to collection conversion triggering is going to help.

It could, because collections have a `zip` operation that takes a collection as parameter. But as you’ve shown, the conversion isn’t triggered and we get compilation errors…

**Comment by dwijnand** [comment]
@julienrf

> It could, because collections have a `zip` operation that takes a collection as parameter.

ah I see what you mean now! yeah implicit conversion will only kick in the method is missing, it won't look for alternative "overloads".

@hrhino 

> Since 2.13 is going to require users to adjust their collections usage anyways, I propose that we can drop this feature.

let me change my PR to propose that, let's see what the team thinks.

**Comment by hrhino** [comment]
To argue against myself: there's a consistency argument that while we can remove `option.zip(list)` we cannot remove `list.zip(option)` as long as there's an `Option[A] => Iterable[A]` implicit conversion, which I can't see going away any time soon.

That said I'd still rather see `Option#zip` only take `Option`s.

File: src/library/scala/Option.scala [hunk 2]
```diff
@@ -8,6 +8,8 @@
 
 package scala
 
+import scala.collection.GenIterable
```
Comment (hrhino) [inline src/library/scala/Option.scala]: unused import

File: src/library/scala/Option.scala [hunk 3]
```diff
@@ -301,6 +303,26 @@ sealed abstract class Option[+A] extends Product with Serializable {
...
+   *  Some("foo") zip None
+   *
+   *  // Returns None because `this` option is empty.
+   *  None zip Some("bar")
+   *  }}}
+   *
+   *  @param  that   the options which is going to be zipped
+   */
+  final def zip[A1 >: A, B](that: Option[B]): Option[(A1, B)] =
+    if (isEmpty || that.isEmpty) None else Some(this.get, that.head)
```
Comment (hrhino) [inline src/library/scala/Option.scala]: `that.get` avoids a wrapping

**Comment by Ichoran** [comment]
Why not have two signatures on `Option`:

```
def zip[B](that: Option[B]): Option([A, B]) = ???
def zip[B, C[B] <: CollectionWithZip[B]](that: C): C[B] = ???
```

I know how to write the second one with CBF.  Not sure I'd do it right from memory for the new collections.

**Comment by SethTisue** [comment]
I agree with @hrhino: I think we should have `Option#zip(Option)` only and not overload it. Because:

* As much as possible, `Option` should be self-contained and not pull in collections stuff.
* `myOption.zip(myCollection.headOption)` is clearer, better, more explicit code anyway.

Dale says he's not working on this anymore, but I (at least) would welcome a new PR on this.

**Comment by jozic** [comment]
this is at least third PR (after #4836 and #4437) with same intent which gets closed. 
@SethTisue can we just reopen this one? what else need to be done here? it looks reasonable as it is, no? 2.13 is going to be non compatible with 2.12 anyway? 
I'm open for picking this up and finishing it

**Comment by dwijnand** [comment]
I've reopened the pr, in case it is mergeable as is.

**Comment by SethTisue** [comment]
@jozic this is nearly to the finish line, but, some of the review feedback at https://github.com/scala/scala/pull/4437 has not been addressed yet

want to do the last tweaks...? (#4437 is the only other ticket you need to look at, I looked at the others and didn't find anything still pending.)

**Comment by jozic** [comment]
thanks @SethTisue, I will check it out soon

**Comment by dwijnand** [comment]
@SethTisue Please let me know if I missed any review feedback at #4437.

**Review by SethTisue — APPROVED** [review]
I think this is good to go. (We should squash one last time before merge.)

**Comment by jozic** [comment]
thank you guys! Hopefully this time we gonna get it!

**Comment by dwijnand** [comment]
Rebased and squashed.

**Comment by SethTisue** [comment]
thanks everybody, on this group effort

**Pull request merged**